### PR TITLE
Rreg fpu changes

### PR DESCRIPTION
**Pull request by skuater**
- implemented types ut80 ut96 ut128 and utX
- refactorized all r_debug_reg_sync call with R_REG_TYPE_XXX and true/false
- implemented reg list for values ut80, ut96 and ut128 into r_debug_reg_list
- command 'dr' and 'ar' now can show ut80 ut96 and ut128 registers (TODO set regs >64 with dr command)
- command 'arb' now can show distinct arenas.
- r_debug_reg_sync now know how to manage the registers redirection (@ token into profile)
- r_debug_reg_get_err ,now have a new parameter to get a utX struct with the values >64 bits, err return the numbers of bits(80, 96, 128) , 0 for x64 values and the value is the return, or 1 if error.
- refactorized all code referencing regs arenas to implement the @ token into profile
- implemented AVX register access under windows
- command 'drf' on windows show debug information about fpu and AVX registers values
- refactorized the x86 windows debug profile with the new @ token.
- some other fix i dont remember

### CODE REVIEW

File: libr/reg/value.c [hunk 1]
```diff
@@ -3,14 +3,52 @@
 #include <r_reg.h>
 #include <r_util.h>
 
+R_API ut64 r_reg_get_value_big(RReg *reg, RRegItem *item, utX *val) {
+	RRegSet *regset;
+	int off;
+	ut64 ret = 0LL;
+	if (!reg || !item)
```
Comment (radare) [inline libr/reg/value.c]: missing braces {}

File: libr/debug/p/native/w32.c [hunk 2]
```diff
@@ -824,6 +857,205 @@ void w32_break_process (void *d) {
...
+
+static int w32_reg_write (RDebug *dbg, int type, const ut8* buf, int size) {
+	BOOL ret = false;
+	HANDLE thread;
+	CONTEXT ctx __attribute__((aligned (16)));
+	thread = w32_open_thread (dbg->pid, dbg->tid);
+	ctx.ContextFlags = CONTEXT_ALL;
+	GetThreadContext (thread, &ctx);
+	if (type == R_REG_TYPE_GPR) {
+		if (size > sizeof(CONTEXT)) {
```
Comment (radare) [inline libr/debug/p/native/w32.c]: space before (

File: libr/debug/p/native/w32.c [hunk 3]
```diff
@@ -824,6 +857,205 @@ void w32_break_process (void *d) {
...
+			size = sizeof(CONTEXT);
```
Comment (radare) [inline libr/debug/p/native/w32.c]: same

File: libr/debug/p/native/w32.c [hunk 4]
```diff
@@ -824,6 +857,205 @@ void w32_break_process (void *d) {
...
+	for (x = 0; x < 8; x++) {
+		// the conversin from long double to double only work for compilers
+		// with long double size >=10 bytes (also we lost 2 bytes of precision)
+		//   in mingw long double is 12 bytes size
+		//   in msvc long double is alias for double = 8 bytes size
+		//   in gcc long double is 10 bytes (correct representation)
+		eprintf("ST%i %04x %016"PFMT64x" (%f)\n", x, st[x].High, st[x].Low, (double)(*((long double *)&st[x])));
+	}
+	for (x = 0; x < 8; x++) {
+		eprintf("MM%i %016"PFMT64x"\n", x, mm[x]);
```
Comment (radare) [inline libr/debug/p/native/w32.c]: missing space before (

File: libr/debug/p/native/w32.c [hunk 5]
```diff
@@ -824,6 +857,205 @@ void w32_break_process (void *d) {
...
+	}
+	for (x = 0; x < nxmm; x++) {
+		eprintf("XMM%i %016"PFMT64x" %016"PFMT64x"\n", x, xmm[x].High, xmm[x].Low);
+	}
+	// show Ymm regs
+	nymm = GetAVX(hThread, &xmm, &ymm);
```
Comment (radare) [inline libr/debug/p/native/w32.c]: space

File: libr/debug/p/native/w32.c [hunk 6]
```diff
@@ -824,6 +857,205 @@ void w32_break_process (void *d) {
...
+		//   in mingw long double is 12 bytes size
+		//   in msvc long double is alias for double = 8 bytes size
+		//   in gcc long double is 10 bytes (correct representation)
+		eprintf("ST%i %04x %016"PFMT64x" (%f)\n", x, st[x].High, st[x].Low, (double)(*((long double *)&st[x])));
+	}
+	for (x = 0; x < 8; x++) {
+		eprintf("MM%i %016"PFMT64x"\n", x, mm[x]);
+	}
+	for (x = 0; x < nxmm; x++) {
+		eprintf("XMM%i %016"PFMT64x" %016"PFMT64x"\n", x, xmm[x].High, xmm[x].Low);
```
Comment (radare) [inline libr/debug/p/native/w32.c]: missing space

File: libr/debug/p/native/w32.c [hunk 7]
```diff
@@ -824,6 +857,205 @@ void w32_break_process (void *d) {
...
+	}
+	Success = w32_InitializeContext(NULL, CONTEXT_ALL | CONTEXT_XSTATE, NULL, &ContextSize);
+	if ((Success == TRUE) || (GetLastError() != ERROR_INSUFFICIENT_BUFFER)) {
+		return 0;
+	}
+	buffer = malloc(ContextSize);
+	if (buffer == NULL) {
+		return 0;
+	}
+	Success = w32_InitializeContext(buffer, CONTEXT_ALL | CONTEXT_XSTATE, &Context, &ContextSize);
```
Comment (radare) [inline libr/debug/p/native/w32.c]: just do if (!w32_InitializeContext()) {...

File: libr/debug/p/native/w32.c [hunk 8]
```diff
@@ -824,6 +857,205 @@ void w32_break_process (void *d) {
...
+	FeatureMask = w32_GetEnabledXStateFeatures();
+	if ((FeatureMask & XSTATE_MASK_AVX) == 0) {
+		return 0;
+	}
+	Success = w32_InitializeContext(NULL, CONTEXT_ALL | CONTEXT_XSTATE, NULL, &ContextSize);
+	if ((Success == TRUE) || (GetLastError() != ERROR_INSUFFICIENT_BUFFER)) {
+		return 0;
+	}
+	buffer = malloc(ContextSize);
+	if (buffer == NULL) {
```
Comment (radare) [inline libr/debug/p/native/w32.c]: use !

File: libr/debug/p/native/w32.c [hunk 9]
```diff
@@ -824,6 +857,205 @@ void w32_break_process (void *d) {
...
+	PCONTEXT Context;
+	if (w32_GetEnabledXStateFeatures == (ut64 WINAPI (*) ())-1) {
+		return 0;
+	}
+	// Check for AVX extension
+	FeatureMask = w32_GetEnabledXStateFeatures();
+	if ((FeatureMask & XSTATE_MASK_AVX) == 0) {
+		return 0;
+	}
+	Success = w32_InitializeContext(NULL, CONTEXT_ALL | CONTEXT_XSTATE, NULL, &ContextSize);
```
Comment (radare) [inline libr/debug/p/native/w32.c]: missing space before (

**Comment by alvarofe** [comment]
This introduces some weird behavior. Besides I had to fix an oob (merged in master). Mind you take a look at it since you are familiar with the code involved?


```
[0x100001174]> pd~20
r_reg_get_value: null or oob arena for current regset
r_reg_get_value: 32bit oob read 144
r_reg_get_value: null or oob arena for current regset
r_reg_get_value: 32bit oob read 144
r_reg_get_value: null or oob arena for current regset
r_reg_get_value: 32bit oob read 144
        │   0x1000011a1      e859320000     call sym.func.1000043ff
        │   0x1000011c8      c705fe420000.  mov dword [0x1000054d0], 0x50 ; 'P' ; [0x1000054d0:4]=80 LEA section.10.__data ; "P" @ 0x1000054d0
      │││   0x1000011eb      e8bc320000     call sym.imp.atoi
     │  │   0x100001200      31c0           xor eax, eax
     │  │   0x100001202      e829330000     call sym.imp.ioctl
     │  │   0x100001207      83f8ff         cmp eax, -1
     │ ┌──< 0x10000120a      740e           je 0x10000121a
     │ ││   0x10000120c      0fb745d2       movzx eax, word [rbp - 0x2e]
     └────> 0x100001214      8905b6420000   mov dword [0x1000054d0], eax ; [0x1000054d0:4]=80 LEA section.10.__data ; "P" @ 0x1000054d0
       │    0x100001230      e8d7320000     call sym.imp.getenv
       ││   0x10000123d      e86a320000     call sym.imp.atoi
       ││   0x100001242      890588420000   mov dword [0x1000054d0], eax ; [0x1000054d0:4]=80 LEA section.10.__data ; "P" @ 0x1000054d0
       └└─> 0x100001248      e8d1320000     call sym.imp.getuid
       │    0x10000126d      c6058c420000.  mov byte [0x100005500], 1  ; [0x100005500:1]=207 LEA section.11.__bss ; secti
```

**Pull request merged**